### ISSUE TITLE
pxd file is not processed when an Extension object is cythonized

### ISSUE THREAD

**Issue by fduxiao** [cython/cython#4372]
**Description**
I am writing a python wrapper for a C++ project which uses the [nlohmann/json](https://github.com/nlohmann/json) library for its config processing. Following this [tutorial](https://cython.readthedocs.io/en/latest/src/userguide/sharing_declarations.html), I first made a wrapper for the nlohmann::json class, which provides the pxd file to be further included by the wrapper for the whole project. I organized the files like this:
```shell
$ tree .
.
├── pkg
│   └── nlohmann
│       ├── interface.hpp
│       ├── nlohmann.pxd
│       └── nlohmann.pyx
└── setup.py
```

If I put
```python3
from setuptools import setup
from Cython.Build import cythonize

extensions = cythonize([
    "./pkg/nlohmann/nlohmann.pyx"
], language_level="3", build_dir="build", language="c++")

extensions[0].include_dirs.append("./pkg/nlohmann")
extensions[0].name = "pkg.nlohmann"

setup(ext_modules=extensions)
```
in `setup.py`, then it works well, (but I have to specify include_dirs and the location with several other lines). 

However, if I use the `Extension` class like this:
```python3
from setuptools import setup
from Cython.Distutils import Extension
from Cython.Build import cythonize

extensions = cythonize([
    Extension(
        "pkg.nlohmann", 
        ["./pkg/nlohmann/nlohmann.pyx"], 
        language="c++",
        include_dirs=["./pkg/nlohmann/"]
    )
], language_level="3", build_dir="build")

setup(ext_modules=extensions)
```
in `setup.py`, then cython will complains about the undefined symbols in `./pkg/nlohmann/nlohmann.pyx`, which have already been defined in `./pkg/nlohmann/nlohmann.pxd`.

**Environment**
OS: 
```shell
$ uname -a
Darwin xiaodeMacBook-Pro.local 20.4.0 Darwin Kernel Version 20.4.0: Fri Mar  5 01:14:14 PST 2021; root:xnu-7195.101.1~3/RELEASE_X86_64 x86_64
```
Python([Homebrew](https://github.com/Homebrew/brew)):
```shell
$ python3 --version
Python 3.9.7
```
Cython:
```shell
$ cython --version 
Cython version 0.29.21
```

**To Reproduce**
If you have installed the nlohmann/json library, then  try the following files.

pkg/nlohmann/interface.hpp:
```c++
#include <nlohmann/json.hpp>
#include <string>
#include <vector>

void set_nil(nlohmann::json& j) 
{
    j = nullptr;
}

void set_bool(nlohmann::json& j, bool b) 
{
    j = b;
}

void set_int(nlohmann::json& j, long long l) 
{
    j = l;
}

void set_string(nlohmann::json& j, std::string s) 
{
    j = s;
}

void set_empty_array(nlohmann::json& j) 
{
    j = nlohmann::json::array();
}

void array_push_back(nlohmann::json& j, nlohmann::json e) 
{
    j.push_back(e);
}

void set_empty_object(nlohmann::json& j) 
{
    j = nlohmann::json::object();
}

void set_object_item(nlohmann::json& j, std::string key, nlohmann::json& value) 
{
    j[key] = value;
}

void set_json(nlohmann::json& j1, nlohmann::json& j2) 
{
    j1 = j2;
}

std::string dump(nlohmann::json& j) 
{
    return j.dump();
}
```

pkg/nlohmann/nlohmann.pxd:
```C++
from libcpp cimport bool as cppbool
from libcpp.vector cimport vector
from libcpp.string cimport string

cdef extern from "<nlohmann/json.hpp>" namespace "nlohmann":
    cdef cppclass json:
        pass

cdef extern from "interface.hpp":
    void set_nil(json &j)
    void set_bool(json &j, cppbool b)
    void set_int(json &j, long long l)
    void set_string(json &j, string s)

    void set_empty_array(json& j)
    void array_push_back(json& j, json e)

    void set_empty_object(json &j)
    void set_object_item(json& j, string key, json& value)

    void set_json(json& j1, json& j2)
    string dump(json& j)


cdef class Json:
    cdef json j
    cdef set_json(self, Json j)
```

pkg/nlohmann/nlohmann.pyx:
```C++
cdef class Json:
    def dump_b(self):
        return dump(self.j)

    def dump_s(self):
        return dump(self.j).decode()

    def __init__(self, x=None):
        self.from_python(x)

    cdef set_json(self, Json j):
        set_json(self.j, j.j)

    def from_python(self, x):
        if x is None:
            set_nil(self.j)
        elif isinstance(x, bool):
            set_bool(self.j, x)
        elif isinstance(x, int):
            set_int(self.j, x);
        elif isinstance(x, str):
            set_string(self.j, x.encode())
        elif isinstance(x, dict):
            set_empty_object(self.j)
            for key, value in x.items():
                set_object_item(self.j, key.encode(), Json(value).j)
        elif isinstance(x, list):
            set_empty_array(self.j)
            for i in x:
                array_push_back(self.j, Json(i).j)
        elif isinstance(x, Json):
            self.set_json(x)
        else:
            raise NotImplementedError(type(x))
```

**Comment by fduxiao** [comment]
I believe the pxd file is not processed because even if I cimport it in the pyx file, cython still complains that I am using the python object `self.j` as a `nlohmann::json` object.

pkg/nlohmann/nlohmann.pyx:
```cython
cimport nlohmann


cdef class Json:
    def dump_b(self):
        return nlohmann.dump(self.j)

    def dump_s(self):
        return nlohmann.dump(self.j).decode()

    def __init__(self, x=None):
        self.from_python(x)

    cdef set_json(self, Json j):
        nlohmann.set_json(self.j, j.j)

    def from_python(self, x):
        if x is None:
            nlohmann.set_nil(self.j)
        elif isinstance(x, bool):
            nlohmann.set_bool(self.j, x)
        elif isinstance(x, int):
            nlohmann.set_int(self.j, x);
        elif isinstance(x, str):
            nlohmann.set_string(self.j, x.encode())
        elif isinstance(x, dict):
            nlohmann.set_empty_object(self.j)
            for key, value in x.items():
                nlohmann.set_object_item(self.j, key.encode(), Json(value).j)
        elif isinstance(x, list):
            nlohmann.set_empty_array(self.j)
            for i in x:
                nlohmann.array_push_back(self.j, Json(i).j)
        elif isinstance(x, Json):
            self.set_json(x)
        else:
            raise NotImplementedError(type(x))

```

**Comment by 0dminnimda** [comment]
It would be really nice if you could squeeze this into a minimal, reproducible example.

**Comment by fduxiao** [comment]
@0dminnimda I squeezed it into this [repo](https://github.com/fduxiao/cythonize_module). 

Or you can read the following directly. 

pkg/nlohmann/interface.hpp:
```C++
#include <nlohmann/json.hpp>
#include <string>
#include <vector>

void set_nil(nlohmann::json& j) {
    j = nullptr;
}
```

pkg/nlohmann/nlohmann.pxd:
```cython
cdef extern from "<nlohmann/json.hpp>" namespace "nlohmann":
    cdef cppclass json:
        pass

cdef extern from "interface.hpp":
    void set_nil(json &j)


cdef class Json:
    cdef json j
```

pkg/nlohmann/nlohmann.pyx:
```cython
cdef class Json:
    def __init__(self, x=None):
        self.from_python(x)

    def from_python(self, x):
        if x is None:
            set_nil(self.j)
        else:
            raise NotImplementedError(type(x))
```

setup.py:
```python
from setuptools import setup
from Cython.Distutils import Extension
from Cython.Build import cythonize

extensions = cythonize([
    Extension(
        "pkg.nlohmann", 
        ["./pkg/nlohmann/nlohmann.pyx"], 
        language="c++",
        include_dirs=["./pkg/nlohmann/"]
    )
], language_level="3", build_dir="build")

setup(ext_modules=extensions)
```

**Comment by 0dminnimda** [comment]
So, we don't need interface.cpp?

**Comment by scoder** [comment]
Please add "__init__.py" files to the package directories to turn them into actual packages.

**Comment by fduxiao** [comment]
@0dminnimda the `set_nil` in interface.cpp is used in nlohmann.pyx.

@scoder `nlohmann` is intended as a module, not a package, and even if I put the __init__.py file in that directory, cython still complains about the same problem.

**Comment by fduxiao** [comment]
@0dminnimda I finally figured out how this happened. It will work if I change the name passed to `Extension` to `pkg.nlohmann.nlohmann`. I intended to use the directory to organize the code related to module `nlohmann`, and build the extension as a module next to the directory, but the pxd files are searched based on the module name instead of source file path ([here](https://github.com/cython/cython/blob/aea4e6b84b38223c540266f8c57093ee2039f284/Cython/Compiler/Pipeline.py#L32)). Or I can move the pxd file next to the place of the module. Thanks for your help.
